### PR TITLE
[NET 6 Attribute Conversion] AVFoundation

### PR DESCRIPTION
**Pull request by chamons**
This is the first non-trivial conversion using [mellite](https://github.com/chamons/mellite), starting with AVFoundation.

It is best reviewed commit by commit.

## Note as you review:

- The NET6 attributes will sometimes pick up what look like "unnecessary" attributes such as:

```patch
-#if !NET
-		[iOS (10,0), Mac (10,12)]
+#if NET
+		[SupportedOSPlatform ("ios10.0")]
+		[SupportedOSPlatform ("macos10.12")]
+		[UnsupportedOSPlatform ("tvos")]
+#else
+		[iOS (10,0)]
+		[Mac (10,12)]
#endif
```

This is because the definition, in this example of AVAudioRecorder, has `[NoTV]` [here](https://github.com/xamarin/xamarin-macios/blob/main/src/avfoundation.cs#L1769-L1772), and we have to bring that down to each decorated API element.

- Because of this, keep a copy of [avfoundation.cs](https://github.com/xamarin/xamarin-macios/blob/main/src/avfoundation.cs) open for quick reference.
- Watch is explicitly ignored, both from the parent and element contexts, so you will not see a SupportedOSPlatform/UnsupportedOSPlatform watch.
- This was generated using convert script found [here](https://github.com/chamons/mellite/blob/master/scripts/convert.sh) `./scripts/convert.sh AVFoundation`
- The ordering is explicitly `#if NET` first upon Rolf's request, so a number of cases were flipped.
- The first two commit comment out defines that are restored in the 5th and 6th commit, as the tool can not process the file otherwise.

# Areas for discussion
- src/AVFoundation/AVCaptureConnection.cs has :
```
		[SupportedOSPlatform ("maccatalyst14.0")]
		[UnsupportedOSPlatform ("ios7.0")]
```
which seems strange, but is explicitly implied by the attributes. Is this expected/desired?
- src/AVFoundation/AVCompat.cs and src/AVFoundation/AVTypes.cs has some obsoletes that were there along side a Deprecated. I was going to remove them, but I could scope them in !NET if desired
- in src/AVFoundation/Events.cs some of the obsolete texts changed case as I generate them from the legacy version. Is that fine?`[Obsolete ("Starting with ios6.0 Use 'AVAudioSession.Notification.ObserveAudioRouteChange' instead.", DiagnosticId = "BI1234", UrlFormat = "https://github.com/xamarin/xamarin-macios/wiki/Obsolete")]`

**I will be running this tool and process many times, so if there are improvements or corrects please speak up.**

### CODE REVIEW

File: src/AVFoundation/AVCompat.cs [hunk 1]
```diff
@@ -217,8 +217,15 @@ public virtual void RespondByRequestingPersistableContentKeyRequest ()
 
 #if TVOS
 	// tvOS removed some types - we need to keep stubs of them for binary compatibility
-	[Obsolete ("Removed in tvOS 10.")]
+#if NET
```
Comment (rolfbjarne) [inline src/AVFoundation/AVCompat.cs:220]: This entire file is inside a !NET, so there's no need to change anything in here:

https://github.com/xamarin/xamarin-macios/blob/c61ba7112545276d5ce818f62d236cefef88e847/src/AVFoundation/AVCompat.cs#L3

File: src/AVFoundation/AVTypes.cs [hunk 2]
```diff
@@ -258,10 +258,17 @@ public static CGRect WithAspectRatio (this CGRect self, CGSize aspectRatio)
...
+	[SupportedOSPlatform ("macos10.10")]
+	[UnsupportedOSPlatform ("ios")]
+	[UnsupportedOSPlatform ("tvos")]
+#else
+	[Mac (10, 10)]
+	[NoiOS]
+	[NoWatch]
+	[NoTV]
 #endif
+	[Obsolete ("This API is not available on this platform.")]
```
Comment (rolfbjarne) [inline src/AVFoundation/AVTypes.cs:271]: This Obsolete should still be inside an `#if !MONOMAC`

File: src/AVFoundation/AVTypes.cs [hunk 3]
```diff
@@ -274,10 +281,17 @@ public struct AVSampleCursorSyncInfo {
...
```
Comment (rolfbjarne) [inline src/AVFoundation/AVTypes.cs:294]: Same.

File: src/AVFoundation/AVTypes.cs [hunk 4]
```diff
@@ -299,20 +313,34 @@ public struct AVSampleCursorDependencyInfo {
...
```
Comment (rolfbjarne) [inline src/AVFoundation/AVTypes.cs:326]: Same.

File: src/AVFoundation/AVTypes.cs [hunk 5]
```diff
@@ -299,20 +313,34 @@ public struct AVSampleCursorDependencyInfo {
...
```
Comment (rolfbjarne) [inline src/AVFoundation/AVTypes.cs:343]: Same.

**Review by rolfbjarne — CHANGES_REQUESTED** [review]

File: src/AVFoundation/AVCompat.cs (hunk 1 above)
Comment (dalexsoto) [reply src/AVFoundation/AVCompat.cs:220]: Nice catch

Comment (chamons) [reply src/AVFoundation/AVCompat.cs]: Corrected.

File: src/AVFoundation/AVCaptureConnection.cs [hunk 6]
```diff
@@ -37,10 +37,15 @@
...
-#if !NET
-		[Deprecated (PlatformName.iOS, 7, 0)]
-#else
+#if NET
+		[SupportedOSPlatform ("maccatalyst14.0")]
 		[UnsupportedOSPlatform ("ios7.0")]
+#if IOS
+		[Obsolete ("Starting with ios7.0", DiagnosticId = "BI1234", UrlFormat = "https://github.com/xamarin/xamarin-macios/wiki/Obsolete")]
+#endif
+		[UnsupportedOSPlatform ("tvos")]
```
Comment (tj-devel709) [inline src/AVFoundation/AVCaptureConnection.cs:46]: Personal preference, but I think this line should go up 2 lines with the other 'if NET' supported and not supported (happens in a few other places as well)

**Review by tj-devel709 — APPROVED** [review]

File: src/AVFoundation/AVCaptureConnection.cs (hunk 6 above)
Comment (chamons) [reply src/AVFoundation/AVCaptureConnection.cs:46]: Right now the generator tries to keep like block near each other, and considers:

```
	[UnsupportedOSPlatform ("ios7.0")]
#if IOS
	[Obsolete ("Starting with ios7.0", DiagnosticId = "BI1234", UrlFormat = "https://github.com/xamarin/xamarin-macios/wiki/Obsolete")]
#endif
```

to be one block, as they are both the output of Deprecated on iOS.

**Comment by chamons** [comment]
This PR also hit typo/duplicate issues. 2/4 PRs hit that, so future PRs will run that before submission.

**Comment by chamons** [comment]
Other test will fail until https://github.com/xamarin/xamarin-macios/pull/13745/files lands and is merged in.

**Review by rolfbjarne — APPROVED** [review]

File: src/AVFoundation/Enums.cs [hunk 7]
```diff
@@ -706,84 +998,185 @@ public enum AVContentAuthorizationStatus : long {
...
 	[Native]
 	public enum AVPlayerViewControllerSkippingBehavior : long {
 		Default = 0,
 		SkipItem
 	}
 
-	[TV (10,2), Mac (10,12,4), iOS (10,3), Watch (7,0)]
+#if NET
+	[SupportedOSPlatform ("tvos10.2")]
+	[SupportedOSPlatform ("macos12.4")]
```
Comment (tj-devel709) [inline src/AVFoundation/Enums.cs:1172]: Should this be MacOS 10.12.4? (I think the min supported is MacOS 10.14)

File: src/AVFoundation/Enums.cs [hunk 8]
```diff
@@ -794,7 +1187,16 @@ public enum AVContentKeyRequestStatus : long {
 		Failed
 	}
 
-	[TV (10, 2), Mac (10, 12, 4), iOS (10, 3), Watch (7, 0)]
+#if NET
+	[SupportedOSPlatform ("tvos10.2")]
+	[SupportedOSPlatform ("macos12.4")]
```
Comment (tj-devel709) [inline src/AVFoundation/Enums.cs:1192]: same here

File: src/AVFoundation/Enums.cs [hunk 9]
```diff
@@ -804,34 +1206,75 @@ public enum AVContentKeyRequestRetryReason {
 		ReceivedObsoleteContentKey,
 	}
 
-	[TV (10, 2), Mac (10, 12, 4), iOS (10, 3), Watch (7, 0)]
+#if NET
+	[SupportedOSPlatform ("tvos10.2")]
+	[SupportedOSPlatform ("macos12.4")]
```
Comment (tj-devel709) [inline src/AVFoundation/Enums.cs:1211]: Here as well

File: src/AVFoundation/Enums.cs [hunk 10]
```diff
@@ -1168,63 +1961,158 @@ public enum AVFileTypeProfile {
...
 		Playback = 0,
 		PlayAndRecord = 1,
 		PlayAndRecordVoice = 2,
 	}
 
 #if !WATCH
-	[TV (10,2), Mac (10,12,4), iOS (10,3), NoWatch]
+#if NET
+	[SupportedOSPlatform ("tvos10.2")]
+	[SupportedOSPlatform ("macos12.4")]
```
Comment (tj-devel709) [inline src/AVFoundation/Enums.cs:1986]: Same

File: src/AVFoundation/Enums.cs (hunk 7 above)
Comment (chamons) [reply src/AVFoundation/Enums.cs:1172]: Yes, I just caught this in my AppKit work locally. I'll have to fix the generator and re-run here.

**Comment by chamons** [comment]
Failing test was:
```
+Mac Catalyst [dotnet] (TimedOut; Succeeded)
```

I'm comfortable landing this.

**Comment by chamons** [comment]
https://github.com/xamarin/maccore/issues/2414

**Pull request merged**